cli/azd/test/ostest/ostest.go

```diff
@@ -1,157 +1,35 @@
-	temp := t.TempDir()
-
-	if runtime.GOOS == "windows" {
-		// Enable our additional custom remove logic for Windows where we see locked files.
-		t.Cleanup(func() {
-			err := removeAllWithDiagnostics(t, temp)
-			if err != nil {
-				logHandles(t, temp)
-				t.Fatalf("TempDirWithDiagnostics: %s", err)
-			}
-		})
-	}
-
-	return temp
-}
-
-func logHandles(t *testing.T, path string) {
-	handle, err := exec.LookPath("handle")
-	if err != nil && errors.Is(err, exec.ErrNotFound) {
-		t.Logf("handle.exe not present. Skipping handle detection. PATH: %s", os.Getenv("PATH"))
-		return
-	}
-
-	if err != nil {
-		t.Logf("failed to find handle.exe: %s", err)
-		return
-	}
-
-	args := azdexec.NewRunArgs(handle, path, "-nobanner")
-	cmd := azdexec.NewCommandRunner()
-	rr, err := cmd.Run(context.Background(), args)
-	if err != nil {
-		t.Logf("handle.exe failed. stdout: %s, stderr: %s\n", rr.Stdout, rr.Stderr)
-		return
-	}
-
-	t.Logf("handle.exe output:\n%s\n", rr.Stdout)
-
-	// Ensure telemetry is initialized since we're running in a CI environment
-	_ = telemetry.GetTelemetrySystem()
-
-	// Log this to telemetry for ease of correlation
-	tracer := telemetry.GetTracer()
-	_, span := tracer.Start(context.Background(), "test.file_cleanup_failure")
-	span.SetAttributes(attribute.String("handle.stdout", rr.Stdout))
-	span.SetAttributes(attribute.String("ci.build.number", os.Getenv("BUILD_BUILDNUMBER")))
-	span.End()
-}
-
-func removeAllWithDiagnostics(t *testing.T, path string) error {
-	retryCount := 0
-	loggedOnce := false
-	return retry.Do(context.Background(), retry.WithMaxRetries(10, retry.NewConstant(1*time.Second)), func(_ context.Context) error {
-		removeErr := os.RemoveAll(path)
-		if removeErr == nil {
-			return nil
-		}
-		t.Logf("failed to clean up %s with error: %v", path, removeErr)
-
-		if retryCount >= 2 && !loggedOnce {
-			// Only log once after 2 seconds - logHandles is pretty expensive and slow
-			logHandles(t, path)
-			loggedOnce = true
-		}
-
-		retryCount++
-		return retry.RetryableError(removeErr)
-	})
-}
-
-// Setenv sets the value of the environment variable named by the key.
-// Any set values are automatically restored during test Cleanup.
-func Setenv(t *testing.T, key string, value string) {
-	t.Setenv(key, value)
-}
-
 // Unsetenv unsets the environment variable, which is later restored during test Cleanup.
 func Unsetenv(t *testing.T, key string) {
-	orig, present := os.LookupEnv(key)
+	// Call t.Setenv here so that if this test is being run in parallel we will fail.
+	t.Setenv(key, "")
+	// No need to for an explicit t.Cleanup, the call to Setenv above registered a Cleanup function
+	// that will restore the value it had before Unsetenv was called.
 	os.Unsetenv(key)
-
-	t.Cleanup(func() {
-		if present {
-			os.Setenv(key, orig)
-		}
-	})
 }
 
 // Unsetenvs unsets the provided environment variables, which is later restored during test Cleanup.
 func Unsetenvs(t *testing.T, keys []string) {
```

Comment:
I definitely wrote most of this before finding out that go1.18 introduced `t.Cleanup`, but the implementation here was done to avoid introducing `n` cleanup functions for `n` variables to minimize pollution. What was definitely missing was a guard against parallel execution.

Reply:
Yeah - I figured that the set of cleanup functions would be small enough that we didn't need to worry about registering just a single cleanup function.  It's unfortunate that we need to call `t.Setenv` at least once here to get the nice `Parallel` behavior.